File: kraken/src/main/java/glitch/kraken/services/SearchService.java
package glitch.kraken.services;

import glitch.kraken.GlitchKraken;
import glitch.kraken.services.request.ChannelSearchRequest;
import glitch.kraken.services.request.GamesSearchRequest;
import glitch.kraken.services.request.StreamSearchRequest;
import glitch.service.AbstractHttpService;

public class SearchService extends AbstractHttpService {
    public SearchService(GlitchKraken rest) {
        super(rest.getClient(), rest.getHttpClient());
    }

    public ChannelSearchRequest searchChannel(String query) {
        return new ChannelSearchRequest(http, query);
    }

    public GamesSearchRequest searchGames(String query) {
        return new GamesSearchRequest(http, query);
    }

    public StreamSearchRequest searchStreamss(String query) {
        return new StreamSearchRequest(http, query);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/SearchService.kt
package io.glitchlib.v5.service

import io.glitchlib.GlitchClient
import io.glitchlib.internal.http.bodyFlowable
import io.glitchlib.model.OrdinalList
import io.glitchlib.v5.internal.AbstractKrakenService
import io.glitchlib.v5.model.json.Channel
import io.glitchlib.v5.model.json.Game
import io.glitchlib.v5.model.json.Stream
import io.glitchlib.v5.service.request.ChannelSearchRequest
import io.glitchlib.v5.service.request.GamesSearchRequest
import io.glitchlib.v5.service.request.StreamSearchRequest

class SearchService(client: GlitchClient) : AbstractKrakenService(client) {

    fun searchChannel(query: String, request: ChannelSearchRequest.() -> Unit = {}) =
        get<OrdinalList<Channel>>("/search/channels", ChannelSearchRequest(query).apply(request)()).bodyFlowable

    fun searchGames(query: String, request: GamesSearchRequest.() -> Unit = {}) =
        get<OrdinalList<Game>>("/search/games", GamesSearchRequest(query).apply(request)()).bodyFlowable

    fun searchStreams(query: String, request: StreamSearchRequest.() -> Unit = {}) =
        get<OrdinalList<Stream>>("/search/streams", StreamSearchRequest(query).apply(request)()).bodyFlowable
}
